plat/drivers/gic/gic-v2.c

```diff
@@ -509,6 +510,71 @@ static int gicv2_do_probe(const void *fdt)
 
 	/* Set driver functions */
 	gicv2_drv.ops = drv_ops;
+}
+
+#if defined(CONFIG_UKPLAT_ACPI)
+/* Default page-aligned up size for GICv2 CPU interface according to spec */
+#define GICV2_GICC_MEM_SZ	0x2000
+
+static int acpi_get_gicc(struct _gic_dev *g)
+{
+	union {
+		acpi_madt_gicc_t *gicc;
+		acpi_subsdt_hdr_t *h;
+	} m;
+	acpi_madt_t *madt;
+	__sz off, len;
+
+	madt = acpi_get_madt();
+	UK_ASSERT(madt);
+
+	/* In ACPI all GICCs' base address must be the same */
+	len = madt->hdr.tab_len - sizeof(*madt);
+	for (off = 0; off < len; off += m.h->len) {
+		m.h = (acpi_subsdt_hdr_t *)(madt->entries + off);
+
+		if (m.h->type != ACPI_MADT_GICC)
+			continue;
+
+		/* If GICv3/4 this field is 0 */
+		if (!m.gicc->paddr)
+			return -ENOTSUP;
+
+		g->cpuif_mem_addr = m.gicc->paddr;
+		g->cpuif_mem_size = GICV2_GICC_MEM_SZ;
+
+		return 0;
+	}
+
+	return -ENOENT;
+}
+
+static int gicv2_do_probe(void)
+{
+	int rc;
+
+	rc = acpi_get_gicc(&gicv2_drv);
+	if (unlikely(rc < 0))
+		return rc;
+
+	rc = acpi_get_gicd(&gicv2_drv);
+	if (unlikely(rc < 0))
+		return rc;
+
+	if (unlikely(gicv2_drv.dist_mem_size != GICD_V2_MEM_SZ))
```

Comment:
This is because `GICC` entries exist for `GICv3` as well, so `GICv2`'s `acpi_get_gicc` may find `GICC`'s but they are not for the `GICv2`. Now that I take a look at it, since the `GICD` is common among both version, it should be checked for in `plat/drivers/gic/gic-common.c` and then decide upon which `GIC` version to actually probe. However, this cannot be done because `Devicetree` probing must also be accommodated for.... unless we assume `GICD` register map sizes between versions, which I would rather not do. The whole initial `_dtb_init_gic` looks dumb to me.

So, since the `GICD` is common for both, I would make the `gicd` pointer in `acpi_get_gicd` a `static` pointer so that we do not have to search for it a second time if both interrupt controllers end up being enumerated. Furthermore, I would make sure that `acpi_get_gicd` is called first in the `ACPI` probing. Let me know your thoughts.

Reply:
> This is because GICC entries exist for GICv3 as well, so GICv2's acpi_get_gicc may find GICC's but they are not for the GICv2

Can you have a system with both a GICv2 and a GICv3? I haven't see one so far.

> Why aren't LIBGICV2 and LIBGICV3 mutually exclusive?

It's not because you can have a system with > 1 distributors. It's there because it allows a single build to be tested, released, and deployed for different machine types, that use GICv2 and GICv3.

Reply:
> Can you have a system with both a GICv2 and a GICv3? I haven't see one so far.

Not to my knowledge, that is why I asked why aren't the two configs mutually exclusive. What I meant is that if both are enabled, the GICv2 is probed first and it will look for GICC entries, which do not give you the GIC version, so the ACPI probe of GICv2 cannot know the version until the GICD is probed. That is why I said that I should have probably checked for the GICD first and also cache its result in a function locally declared `static` variable, so that when GICv3 is probed, it can fetch the result from when GICv2 was probed.

> It's not because you can have a system with > 1 distributors. It's there because it allows a single build to be tested, released, and deployed for different machine types, that use GICv2 and GICv3.

Ah, yes, it makes sense then. However, what drivers are (de)selected should not be reflected in the core codebase, this should have been transparently probed for in a generic, properly written, platform bus driver. But it's a discussion for later.

> So, since the GICD is common for both, I would make the gicd pointer in acpi_get_gicd a static pointer so that we do not have to search for it a second time if both interrupt controllers end up being enumerated. Furthermore, I would make sure that acpi_get_gicd is called first in the ACPI probing.

Unless you have any objections, I would proceed as I previously suggested.